### DISCUSSION TITLE
Need help with updating plotly figure

### DISCUSSION THREAD

**Discussion by Nachsh** [rio-labs/rio#159] [Q&A] [answered]
I created a simple page with a pyplot.express plot and a button.
I want the plot to be updated with new data upon button press, but the plot is not updated.

class IMM_GUI_PAGE(rio.Component):        
    
    data: pd.DataFrame = pd.DataFrame({'y':[0,1,2,3]})
    print(data)

    def update_plotly_fig(self):
        self.data['y'] = np.random.randint(0,4,4)
        print(self.data)
        
    # on_btn_press()

    def build(self) -> rio.Component:

        # self.on_btn_press()

        fig = px.line(
            self.data,
            height=200,
            width=200,
        )

        # Set x-axis labels to horizontal and remove labels
        fig.update_xaxes(tickangle=0, title_text="")
        fig.update_yaxes(title_text="")

        # strip down the rest of the plot
        fig.update_layout(
            showlegend=False,
            margin=dict(t=10, l=10, b=10, r=10),
        )

   
        return rio.Column(

            rio.Text("My App", style="heading2"),

            rio.Text(
                "This is a sample page. Replace it with your own content."
            ),

            rio.Button(
                "Click me!",
                on_press=self.update_plotly_fig,
            ),

            rio.Plot(
                figure = fig,
                min_height=20,
                min_width=10,
                background=self.session.theme.neutral_color,
            ),

            spacing=2,
            margin=2,
            align_x=0.5,
            align_y=0,
        )

**Comment by Aran-Fey** [comment, marked as answer]
Rio components only update automatically if you assign a new value to one of the component's attributes. There's no way for Rio to know that you've changed the data in your dataframe, so you must explicitly tell it to rebuild your component by calling `self.force_refresh()`:

```python
async def update_plotly_fig(self):
        self.data['y'] = np.random.randint(0,4,4)
        await self.force_refresh()
```

**Reply by Nachsh** [reply]
Thanks @Aran-Fey ! 
For the solution and explanation. I understand better now.
